doc/SAI-Proposal-HW-Resource.md

```diff
@@ -0,0 +1,167 @@
+    For trap group created in step 1.
+3. create_policer()
+4. Attach policer to the trap using set of 
+    SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER attribute.
+```
+
+### Sample New Workflow
+-------------------
+New workflow will identify if a given trap is applicable to ingress/egress/both. This will be done based on the enhanced capability query API.
+```
+1. Query the HW capability for a given trap
+    sai_status_t sai_query_attribute_capability(
+        switch_id,
+        SAI_OBJECT_TYPE_HOSTIF_TRAP,
+        SAI_HOSTIF_TRAP_TYPE_DNAT_MISS,
+        attr_capability);
+
+    SAI adapter will fill attr_capability->attr_stage with appropriate HW pipeline stage where this trap is handled. Lets say give HW supports this attribute in ingress pipeline and will return SAI_ATTR_STAGE_INGRESS as the attr_stage value.
+
+2. Create a policer pool for ingress stage
+    sai_object_id_t ingress_policer_id;
+    sai_attribute_t ingress_policer_attrs[2];
+
+    ingress_policer_attrs[0].id =
+        (sai_attr_id_t)SAI_POLICER_ATTR_METER_TYPE;
+    ingress_policer_attrs[0].value =
+        SAI_METER_TYPE_PACKETS;
+
+    ingress_policer_attrs[1].id =
+        (sai_attr_id_t)SAI_POLICER_ATTR_OBJECT_STAGE;
+    ingress_policer_attrs[1].value =
+        SAI_OBJECT_STAGE_INGRESS;
+
+    sai_hostif_trap_api->create_policer(
+        &ingress_policer_id,
+        2,
+        ingress_policer_attrs);
+
+3. Create a hostif trap group for ingress stage
+    sai_object_id_t trap_group_id;
+    sai_attribute_t trap_group_attrs[3];
+
+    trap_group_attrs[0].id =
+        (sai_attr_id_t)SAI_HOSTIF_TRAP_GROUP_ATTR_QUEUE;
+    trap_group_attrs[0].value.u32 =
+        0x10;
+
+    trap_group_attrs[1].id =
+        (sai_attr_id_t)SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER,
+    trap_group_attrs[1].value.oid =
+        ingress_policer_id;
+
+    trap_group_attrs[2].id =
+        (sai_attr_id_t)SAI_HOSTIF_TRAP_GROUP_ATTR_OBJECT_STAGE,
+    trap_group_attrs[2].value =
+        SAI_OBJECT_STAGE_INGRESS;
+
+    sai_hostif_trap_group_api->create_hostif_trap_group(
+        &trap_group_id,
+        3,
+        trap_group_attrs);
+
+4. Create hostif trap for ingress stage
+    sai_object_id_t trap_id;
+    sai_attribute_t trap_attrs[2];
+
+    trap_attrs[0].id =
+        (sai_attr_id_t) SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE;
+    trap_attrs[0].value =
+        SAI_HOSTIF_TRAP_TYPE_DNAT_MISS;
+
+    trap_attrs[1].id =
+        (sai_attr_id_t)SAI_HOSTIF_TRAP_ATTR_TRAP_GROUP;
+    trap_attrs[1].value.oid =
+        trap_group_id;
+
+    sai_hostif_trap_api->create_hostif_trap(
+        &trap_id,
+        2,
+        trap_attrs);
+```
+
+### Error Handling
+--------------
+New error codepoint SAI_STATUS_MISMATCH is introduced. SAI adapter must 
+return this error status if there is a mismatch between trap group, trap 
+and policer pool.
+- If trap group and trap type is of different stage e.g. trap group is of 
+  SAI_OBJECT_STAGE_INGRESS and trap capability is SAI_ATTR_STAGE_EGRESS

```

Comment:
SAI_OBJECT_STAGE_INGRESS and trap capability is **SAI_OBJECT_STAGE_EGRESS**

Reply:
Done. Good catch

Reply:
Done